models/stock_picking.py

```diff
@@ -61,68 +62,177 @@ def is_compatible_package(self, package_name):
+        """ Validate the conditions for perfoming a swap of the
+            specified packages by considering the picking instance
+            (expects a singleton picking) and relevant move lines.
+
+            Return the move lines related to the expected package,
+            in case a swap is performed, or the ones related to the
+            scanned package, if both packages belong to the same
+            batch.
 
+            Raise a ValidationError in case packages cannot be found
+            in the picking or if the conditions for swapping are not
+            met.
+            
+        """
         self.ensure_one()
 
-        # TODO: not handling swap packages yet
-        raise ValidationError(
-                _('Package %s not found in the operations of %s') %
-                (package, self.name))
+        if not self.picking_type_id.u_allow_swapping_packages:
+            raise ValidationError(
+                _("Cannot swap packages of picking type '%s'")
+                % self.picking_type_id.name)
+
+        exp_pack_mls = self.move_line_ids.filtered(
+            lambda ml: ml.package_id == expected_package)
 
-        allow_swap = self.picking_type_id.u_allow_swapping_packages
-        if not allow_swap:
+        if not exp_pack_mls:
             raise ValidationError(
-                    _('Package %s not found in the operations of %s') %
-                    (package, self.name))
+                _("Expected package cannot be found in picking %s") %
+                self.name)
 
-        expected_package = self.env.context.get('expected_package')
-        if not expected_package:
+        if not scanned_package.has_same_content(expected_package):
             raise ValidationError(
-                    _("Expected package to scan missing."))
+                _("The contents of %s does not match what you have been "
+                  "asked to pick.") % expected_package.name)
 
-        expected_package = Package.get_package(expected_package_name)
-        exp_pack_mls = self.move_line_ids.filtered(lambda ml: ml.package_id == expected_package)
-        if not exp_pack_mls:
+        if scanned_package.location_id != expected_package.location_id:
             raise ValidationError(
-                    _("Expected package cannot be found in picking %s") %
-                    self.name)
+                _("Packages are in different locations and cannot be swapped"))
+
+        # @todo: implement the batch check once we enable multiple
+        # packages for a picking; if the packages are in the same
+        # batch then no-op (check prior art)
+
+        scanned_pack_mls = None
+
+        if scanned_package.is_reserved():
+            scanned_pack_mls = scanned_package.find_move_lines(
+                [('qty_done', '=', 0)])
+
+            if scanned_pack_mls:
+                # We know that all the move lines have the same picking id
+                ml = scanned_pack_mls[0]
```

Comment:
this can be mls_picking = scanned_pack_mls[0].picking_id and use it in 135 and 141